### ISSUE TITLE
Biomarker count by phase doesn't match Plone version

### ISSUE THREAD

**Issue by nutjob4life** [EDRN/P5#214]
In Message-ID: `<4BA81A19-BFC4-4582-96C1-1F4430FD2636@jpl.nasa.gov>`, @dcrichto1 wrote:

> In the pie chart on the portal under biomarkers, how do you calculate biomarker count by phase?  At one point, we had listed the count of the number of biomarkers.  We had worked to have that match the number Sudhir was reporting to the BSA and DCP leadership.   Can we get that backup at some point? I just want to make sure we validate...

**Comment by nutjob4life** [comment]
@dcrichto1 and @mryan4 and @hoodriverheather: 

The [dev version of Portal 6.0.2](https://edrn-dev.jpl.nasa.gov/portal/renaissance/data-and-resources/biomarkers/) (JPL VPN required) shows the following breakdown by count in both the pie chart and the summary statistics beneath the charts:

| Phase | Count |
|:-----:|------:|
|   1   |  1381 |
|   2   |  1264 |
|   3   |   669 |
|   4   |     7 |
|   5   |     1 |
| Total |  3322 |

The [older Plone-based version 5.1.5](https://edrn-dev.jpl.nasa.gov/portal/5/dev/data-and-resources/biomarkers) (JPL VPN required) had this breakdown:

| Phase | Count |
|:-----:|------:|
|   1   |  1674 |
|   2   |   837 |
|   3   |   308 |
|   4   |     5 |
|   5   |     1 |
| Total |  2825 |

Note that the Plone version hasn't ingested from BMDB in quite some time. If I update with more recent data, we get:

| Phase | Count |
|:-----:|------:|
|   1   |  1844 |
|   2   |   922 |
|   3   |   335 |
|   4   |     5 |
|   5   |     1 |
| Total |  3107 |

Which is correct? 

The older Plone computation was essentially (in pseudo-code):

    for each organ in {Lung, Ovary, Pancreas, Colon, etc.}:
        SELECT count(phase), phase FROM organ_data where the organ name matches
        save a mapping of organ[organ name][phase] = count

This gives us a structure like this:

| Organ | Phase | Count |
|:------|:-----:|------:|
| Colon |   1   |     4 |
| Colon |   2   |    20 |
| Colon |   3   |     6 |
| Colon |   4   |     0 |
| Colon |   5   |     0 |
| Ovary |   1   |    50 |
| Ovary |   2   |    78 |
| Ovary |   3   |   101 |
| Ovary |   4   |     2 |
| Ovary |   5   |     1 |
| …     | …     | …     |


So the older Plone portal was only considering phases at the biomarker—organ level.

However, the Plone version then makes the totals "cascade" in each phase in each organ; that is, a "biomarker" counted in phase 5 also is counted in phases 4, 3, 2, and 1 as well. A biomarker in phase 4 is counted in phases 3, 2, and 1 as well. And so on. The algorithm is this:
```
for each organ:
    let c ← count in phase 5 for that organ
    let c ← count in phase 4 ← count in phase 4 + c
    let c ← count in phase 3 ← count in phase 3 + c
    let c ← count in phase 2 ← count in phase 2 + c
    let count in phase 1 ← count in phase 2 + c
```
Only there's an _error in this computation_. Can you spot it? It's in the last line, which should read:

    let count in phase 1 ← count in phase 1 + c

So instead of cascading the count from phase-2 and summing into phase-1, it wipes out phase-1's count with a doubled phase 2.

For example, for Colon, the counts of phases starts with `[4, 20, 6, 0, 0]` and should become `[30, 26, 6, 0 0]` but instead becomes `[52, 26, 6, 0, 0]`. For Ovary, the counts of phases starts `[50, 78, 101, 2, 1]` and should become `[232, 182, 104, 3, 1]` but instead becomes `[364, 182, 104, 3, 1]`.

Now the newer Wagtail version is doing it differently. It acknowledges the definition of a biomarker as a triple between biomarker—organ—protocol, and that phases exist at the protocol level **and** the organ level (and that's why, despite the doubling of phase-2-into-phase-1 in the Plone version, the counts in the Wagtail version are still higher).

It's doing this:
```
p1 ← p2 ← p3 ← p4 ← p5 ← 0
for each biomarker:
    find all the phases of that biomarker at the organ level:
        find all the phases of that organ at the protocol level:
            combine the phase 1 counts at the biomarker level and organ level
            combine the phase 2 counts at the biomarker level and organ level
            combine the phase 3 counts at the biomarker level and organ level
            combine the phase 4 counts at the biomarker level and organ level
            combine the phase 5 counts at the biomarker level and organ level

            p1 ← p1 + the count of all the combined counts at phase 1
            p2 ← p2 + the count of all the combined counts at phase 2
            p3 ← p3 + the count of all the combined counts at phase 3
            p4 ← p4 + the count of all the combined counts at phase 4
            p5 ← p5 + the count of all the combined counts at phase 5
```
Now we chart `p1`, `p2`, `p3`, `p4`, `p5`, and make the summary counts.

But this approach differs from the Plone version in that it doesn't do the "cascade" step. That is, if a biomarker—organ—protocol like "KLK6—Ovary—PLCO Ovarian Phase III Validation Study" says "I'm good for phase 3", then we should count it as "good for phase 2" **and** "good for phase 1".

Or if a biomarker-organ like "Overa—Ovary" (with no protocol) says "I'm good for phase 5", then we should count it as "good for phase 4" and "good for phase 3" and "good for phase 2" and "good for phase 1".

So that's something we need to do.

**However**, this then pleads the question: suppose a biomarker XYZZY has organ Colon at phase 2 ("XYZZY—Colon") but also a protocol for the Colon called "Reification of Poly-Methly Groups in ad hoc Tissues" at phase 3 ("XYZZY—Colon—Reification of Poly-Methyl Groups")?

Does the phase 3 at the biomarker—organ—protocol level _override_ the phase 2 at the biomarker-organ level?

Or, suppose the biomarker-organ level has a higher phase? Suppose it's "XYZZY—Colon" is phase 3 but "XYZZY—Colon—Reification of Poly-Methyl Groups" at phase 2?

Does the later phase in the biomarker-organ level _override_ the earlier phase in the biomarker—organ—protocol level?

Right now, the Wagtail version is summing _both_ and treating them as distinct, hence the higher counts.

That is:

- If XYZZY—Colon is phase 3 and XYZZY—Colon—Reification of Poly-Methyl Groups is phase 2, then to the new portal it's +1 for phase 2 and +1 for phase 3
- If XYZZY—Colon is phase 2 and XYZZY—Colon—Reification of Poly-Methyl Groups is phase 3, then to the new portal it's _also_ +1 for phase 2 and +1 for phase 3

And if we do the cascade step, then it's:

- If XYZZY—Colon is phase 3 and XYZZY—Colon—Reification of Poly-Methyl Groups is phase 2, then to the new portal it's +1 for phase 3, +2 for phase 2, and +3 for phase 1
- If XYZZY—Colon is phase 2 and XYZZY—Colon—Reification of Poly-Methyl Groups is phase 3, then to the new portal it's _also_ +1 for phase 3, +2 for phase 2, and +3 for phase 1

But perhaps the right approach is to just take the _higher phase_ and also _cascade_. If in any biomarker—organ—protocol, if every biomarker—organ phase is higher than every biomarker—organ—protocol phase, then use it. Otherwise, use the biomarker—organ—protocol phase.

What's the right answer?

Personally, I'd love to drop phase at the biomarker—organ level and just have it _live in only one place_.

It would make everything so much simpler!

But then we'd lose our only phase 5 biomarker, Overa! Why? Because while Overa—Overy has phase 5, Overa—Overy—Overa Ovarian Biomarker Panel (Zhen Zhang non-EDRN protocol)—which also have phase 5—cannot exist in the EDRN portal because the protocol doesn't exist—it's not an EDRN protocol! We don't represent the protocol "Overa—Overy—Overa Ovarian Biomarker Panel (Zhen Zhang non-EDRN protocol)" at all, and if phase lives only at the biomarker—organ level, then there's no phase at all for this biomarker.

😱

Help?

**Comment by mryan4** [comment]
Wow, a lot to look at here.
I was off yesterday, I will look at this today.  Thanks. 


Maureen Ryan (Contractor)
Division of Gastroenterology & Hepatology 
Department of Medicine
UNC at Chapel Hill





On Jan 15, 2023, at 5:36 PM, Sean Kelly ***@***.***> wrote:


@dcrichto1 <https://github.com/dcrichto1> and @mryan4 <https://github.com/mryan4> and @hoodriverheather <https://github.com/hoodriverheather>:

The dev version of Portal 6.0.2 <https://edrn-dev.jpl.nasa.gov/portal/renaissance/data-and-resources/biomarkers/> (JPL VPN required) shows the following breakdown by count in both the pie chart and the summary statistics beneath the charts:

Phase	Count
1	1381
2	1264
3	669
4	7
5	1
Total	3322
The older Plone-based version 5.1.5 <https://edrn-dev.jpl.nasa.gov/portal/5/dev/data-and-resources/biomarkers> (JPL VPN required) had this breakdown:

Phase	Count
1	1674
2	837
3	308
4	5
5	1
Total	2825
Note that the Plone version hasn't ingested from BMDB in quite some time. If I update with more recent data, we get:

Phase	Count
1	1844
2	922
3	335
4	5
5	1
Total	3107
Which is correct?

The older Plone computation was essentially (in pseudo-code):

for each organ in {Lung, Ovary, Pancreas, Colon, etc.}:
    SELECT count(phase), phase FROM organ_data where the organ name matches
    save a mapping of organ[organ name][phase] = count
This gives us a structure like this:

Organ	Phase	Count
Colon	1	4
Colon	2	20
Colon	3	6
Colon	4	0
Colon	5	0
Ovary	1	50
Ovary	2	78
Ovary	3	101
Ovary	4	2
Ovary	5	1
…	…	…
So the older Plone portal was only considering phases at the biomarker—organ level.

However, the Plone version then makes the totals "cascade" in each phase in each organ; that is, a "biomarker" counted in phase 5 also is counted in phases 4, 3, 2, and 1 as well. A biomarker in phase 4 is counted in phases 3, 2, and 1 as well. And so on. The algorithm is this:

for each organ:
    let c ← count in phase 5 for that organ
    let c ← count in phase 4 ← count in phase 4 + c
    let c ← count in phase 3 ← count in phase 3 + c
    let c ← count in phase 2 ← count in phase 2 + c
    let count in phase 1 ← count in phase 2 + c
Only there's an error in this computation. Can you spot it? It's in the last line, which should read:

let count in phase 1 ← count in phase 1 + c
So instead of cascading the count from phase-2 and summing into phase-1, it wipes out phase-1's count with a doubled phase 2.

For example, for Colon, the counts of phases starts with [4, 20, 6, 0, 0] and should become [30, 26, 6, 0 0] but instead becomes [52, 26, 6, 0, 0]. For Ovary, the counts of phases starts [50, 78, 101, 2, 1] and should become [232, 182, 104, 3, 1] but instead becomes [364, 182, 104, 3, 1].

Now the newer Wagtail version is doing it differently. It acknowledges the definition of a biomarker as a triple between biomarker—organ—protocol, and that phases exist at the protocol level and the organ level (and that's why, despite the doubling of phase-2-into-phase-1 in the Plone version, the counts in the Wagtail version are still higher).

It's doing this:

p1 ← p2 ← p3 ← p4 ← p5 ← 0
for each biomarker:
    find all the phases of that biomarker at the organ level:
        find all the phases of that organ at the protocol level:
            combine the phase 1 counts at the biomarker level and organ level
            combine the phase 2 counts at the biomarker level and organ level
            combine the phase 3 counts at the biomarker level and organ level
            combine the phase 4 counts at the biomarker level and organ level
            combine the phase 5 counts at the biomarker level and organ level

            p1 ← p1 + the count of all the combined counts at phase 1
            p2 ← p2 + the count of all the combined counts at phase 2
            p3 ← p3 + the count of all the combined counts at phase 3
            p4 ← p4 + the count of all the combined counts at phase 4
            p5 ← p5 + the count of all the combined counts at phase 5
Now we chart p1, p2, p3, p4, p5, and make the summary counts.

But this approach differs from the Plone version in that it doesn't do the "cascade" step. That is, if a biomarker—organ—protocol like "KLK6—Ovary—PLCO Ovarian Phase III Validation Study" says "I'm good for phase 3", then we should count it as "good for phase 2" and "good for phase 1".

Or if a biomarker-organ like "Overa—Ovary" (with no protocol) says "I'm good for phase 5", then we should count it as "good for phase 4" and "good for phase 3" and "good for phase 2" and "good for phase 1".

So that's something we need to do.

However, this then pleads the question: suppose a biomarker XYZZY has organ Colon at phase 2 ("XYZZY—Colon") but also a protocol for the Colon called "Reification of Poly-Methly Groups in ad hoc Tissues" at phase 3 ("XYZZY—Colon—Reification of Poly-Methyl Groups")?

Does the phase 3 at the biomarker—organ—protocol level override the phase 2 at the biomarker-organ level?

Or, suppose the biomarker-organ level has a higher phase? Suppose it's "XYZZY—Colon" is phase 3 but "XYZZY—Colon—Reification of Poly-Methyl Groups" at phase 2?

Does the later phase in the biomarker-organ level override the earlier phase in the biomarker—organ—protocol level?

Right now, the Wagtail version is summing both and treating them as distinct, hence the higher counts.

That is:

If XYZZY—Colon is phase 3 and XYZZY—Colon—Reification of Poly-Methyl Groups is phase 2, then to the new portal it's +1 for phase 2 and +1 for phase 3
If XYZZY—Colon is phase 2 and XYZZY—Colon—Reification of Poly-Methyl Groups is phase 3, then to the new portal it's also +1 for phase 2 and +1 for phase 3
And if we do the cascade step, then it's:

If XYZZY—Colon is phase 3 and XYZZY—Colon—Reification of Poly-Methyl Groups is phase 2, then to the new portal it's +1 for phase 3, +2 for phase 2, and +3 for phase 1
If XYZZY—Colon is phase 2 and XYZZY—Colon—Reification of Poly-Methyl Groups is phase 3, then to the new portal it's also +1 for phase 3, +2 for phase 2, and +3 for phase 1
But perhaps the right approach is to just take the higher phase and also cascade. If in any biomarker—organ—protocol, if every biomarker—organ phase is higher than every biomarker—organ—protocol phase, then use it. Otherwise, use the biomarker—organ—protocol phase.

What's the right answer?

Personally, I'd love to drop phase at the biomarker—organ level and just have it live in only one place.

It would make everything so much simpler!

But then we'd lose our only phase 5 biomarker, Overa! Why? Because while Overa—Overy has phase 5, Overa—Overy—Overa Ovarian Biomarker Panel (Zhen Zhang non-EDRN protocol)—which also have phase 5—cannot exist in the EDRN portal because the protocol doesn't exist—it's not an EDRN protocol! We don't represent the protocol "Overa—Overy—Overa Ovarian Biomarker Panel (Zhen Zhang non-EDRN protocol)" at all, and if phase lives only at the biomarker—organ level, then there's no phase at all for this biomarker.

😱

Help?

**Comment by nutjob4life** [comment]
@mryan4 thanks.

I recommend viewing this issue directly on GitHub so you can see all the formatting of the pseudocode and tables. It should help with understanding.

Also pinging @kristenanton on this so we can get her input too.

**Comment by mryan4** [comment]
(I'm in GitHub now)

One immediate issue I see here is that we should NOT be counting the phase at the organ level.  It is redundant, and is probably what is causing the higher 3022 number.

The way BMDB is set up right now, I enter the phase at the protocol level, but I also fill in the organ level.  I believe that way back in the beginning, the organ level phase was OK because there was probably only one protocol per organ.  That is no longer the case, so we should not use this field in any reports.

**Comment by mryan4** [comment]
My previous comment refers to this portion of your entry, Sean:

However, this then pleads the question: suppose a biomarker XYZZY has organ Colon at phase 2 ("XYZZY—Colon") but also a protocol for the Colon called "Reification of Poly-Methly Groups in ad hoc Tissues" at phase 3 ("XYZZY—Colon—Reification of Poly-Methyl Groups")?

Does the phase 3 at the biomarker—organ—protocol level override the phase 2 at the biomarker-organ level?

Or, suppose the biomarker-organ level has a higher phase? Suppose it's "XYZZY—Colon" is phase 3 but "XYZZY—Colon—Reification of Poly-Methyl Groups" at phase 2?

Does the later phase in the biomarker-organ level override the earlier phase in the biomarker—organ—protocol level?

Right now, the Wagtail version is summing both and treating them as distinct, hence the higher counts.

This is Maureen again - the answer is that biomarker-organ-protocol is the correct and only level we should be considering for phase data.  biomarker-organ is not an accurate representation of the data, and should be disregarded.

**Comment by mryan4** [comment]
Regarding the non-EDRN protocols:

But then we'd lose our only phase 5 biomarker, Overa! Why? Because while Overa—Overy has phase 5, Overa—Overy—Overa Ovarian Biomarker Panel (Zhen Zhang non-EDRN protocol)—which also have phase 5—cannot exist in the EDRN portal because the protocol doesn't exist—it's not an EDRN protocol! We don't represent the protocol "Overa—Overy—Overa Ovarian Biomarker Panel (Zhen Zhang non-EDRN protocol)" at all, and if phase lives only at the biomarker—organ level, then there's no phase at all for this biomarker.

This is Maureen: I don't know how to fix this.  Overa, CancerSEEK did not come through the usual curation channels.  I added them to BMBD because we needed to have them represented on the portal, but there is no EDRN protocol for them.  Jackie tried back in 2021 to get Dr. Kinzler to add a protocol for CancerSEEK, but he never did. 

I don't think overriding the RDF ingest and adding them to the portal is the right way to go (is it even possible?). But without a protocol, which the investigators must add to the DMCC secure site themselves, I am not sure how to fix this.

**Comment by nutjob4life** [comment]
Resolved as of ea29a9de9a915f236d35887a72705ca4358c734a but @kristenanton and @mryan4 will have a future-forward look at improving these statistics.

Currently, pie chart, bar chart, and count-cards all use phases at the organ level only, as the older Plone version of the portal did. The numbers on all match precisely.

**Comment by nutjob4life** [comment]
QC label missing but code is frozen and shipped to NCI.
